### PR TITLE
Added Clothing Configuration Methods

### PR DESCRIPTION
**Pull request by zorg93**
Addresses Issue #521

### CODE REVIEW

**Comment by DorCoMaNdO** [comment]
I think "Style" would be a better name than "Clothing", it has a more general meaning, which is fitting considering that on some models it isn't exactly "Clothing".
Also I think that you should be able to fetch/set a list of Components/PedProps as a whole.
Finally have an inherited class, "FreemodeStyle" to cover all the additional features for freemode models, which would be heritage, overlays and facial features (could also include tattoos), however there are no natives to get the current value of a certain facial feature, it has to be stored in the code.

Also if facial features are to be implemented, you might want to invert some, that's something I've done with my creator, to match up with Rockstar's creator, EyeSize, NoseTipLength, CheekDepth, LipThickness and ChinWidth are inverted, say if moving the slider right was supposed to make EyeSize bigger, it would make it smaller, this could obviously be handled by whoever implements it however they like, but I think it's a better option to invert it in the function.

`public enum FacialFeature : int`
`    {`
`        NoseWidth = 0,`
`        NoseBottomHeight = 1,`
`        NoseTipLength = 2,`
`        NoseBirdgeDepth = 3,`
`        NoseTipHeight = 4,`
`        NoseBroken = 5,`
`        BrowHeight = 6,`
`        BrowDepth = 7,`
`        CheekboneHeight = 8,`
`        CheekboneWidth = 9,`
`        CheekDepth = 10,`
`        EyeSize = 11,`
`        LipThickness = 12,`
`        JawWidth = 13,`
`        JawShape = 14,`
`        ChinHeight = 15,`
`        ChinDepth = 16,`
`        ChinWidth = 17,`
`        ChinIndent = 18,`
`        NeckWidth = 19`
`    }`
`        public static void SetFacialFeature(this Ped ped, FacialFeature feature, float value)`
`        {`
`            if (feature == FacialFeature.EyeSize || feature == FacialFeature.NoseTipLength || feature == FacialFeature.CheekDepth || feature == FacialFeature.LipThickness || feature == FacialFeature.ChinWidth) value = -value;`

`            Function.Call(Hash._SET_PED_FACE_FEATURE, ped, (int)feature, value);`
`        }`

To get facial features working you'd have to have an initial headblend:
`Function.Call(Hash.SET_PED_HEAD_BLEND_DATA, ped, 0, 21, 0, 0, 21, 0, 0.5F, 0.5F, 0F, false);`
`while (!Function.Call<bool>(Hash.HAS_PED_HEAD_BLEND_FINISHED)) Script.Wait(1);`
Might as well do it by default for any freemode model that is spawned.

`Function.Call(Hash.SET_PED_HEAD_BLEND_DATA, ped, shapeFirstID, shapeSecondID, shapeThirdID, skinFirstID, skinSecondID, skinThirdID, shapeMix, skinMix, thirdMix, isParent);`
I never used third shape, but I'm assuming it's an override, first and second shape are the main ones, SkinMix and ShapeMix are values from 0F to 1F, deciding which shape it inherits more details from, 0.5F being the middle point. shape and skin values (not the mix) are supposed to be the same, as each corresponds to a different parent, but you can still have different values if you wish.

I don't remember where I got it but these are the values (gender-wise) for shapes:
0-20 - Male
21-41 - Female
42-44 - Male (DLC)
45 - Female (DLC)

If you don't care about the genders (you can create male/male or female/female mix) then
`            if (shapeFirstID < 0) shapeFirstID = 0;`
`            if (shapeFirstID > 45) shapeFirstID = 45;`
`            if (skinFirstID == -1) skinFirstID = shapeFirstID;`
`            if (skinFirstID < 0) skinFirstID = 0;`
`            if (skinFirstID > 45) skinFirstID = 45;`
`            if (shapeSecondID < 0) shapeSecondID = 0;`
`            if (shapeSecondID > 45) shapeSecondID = 45;`
`            if (skinSecondID == -1) skinSecondID = shapeSecondID;`
`            if (skinSecondID < 0) skinSecondID = 0;`
`            if (skinSecondID > 45) skinSecondID = 45;`
`            if (shapeThirdID < 0) shapeThirdID = 0;`
`            if (shapeThirdID > 45) shapeThirdID = 45;`
`            if (skinThirdID == -1) skinThirdID = shapeThirdID;`
`            if (skinThirdID < 0) skinThirdID = 0;`
`            if (skinThirdID > 45) skinThirdID = 45;`
While far from clean, should suffice. If you do care about genders:
`            if (shapeFirstID < 0 || shapeFirstID > 20 && shapeFirstID < 42 || shapeFirstID > 44) shapeFirstID = 0;`
`            if (skinFirstID == -1) skinFirstID = shapeFirstID;`
`            if (skinFirstID < 0 || skinFirstID > 20 && skinFirstID < 42 || skinFirstID > 44) skinFirstID = 0;`
`            if (shapeSecondID < 21 || shapeSecondID > 41 && shapeSecondID < 45 || shapeSecondID > 45) shapeSecondID = 21;`
`            if (skinSecondID == -1) skinSecondID = shapeSecondID;`
`            if (skinSecondID < 21 || skinSecondID > 41 && skinSecondID < 45 || shapeSecondID > 45) skinSecondID = 21;`
First is male second is female.

**Comment by zorg93** [comment]
@DorCoMaNdO I'm not a huge fan of including head blend data considering its only application afaik is for the 2(or 4) freemode characters. If head blend worked for other characters it may be more useful. thoughts @JohnnyCrazy @crosire

**Comment by DorCoMaNdO** [comment]
I don't think there's any harm in including more functions for developers to use/toy with, there'd be a few things you'd need to store in the code/memory instead of using natives to get the current values of but I'm pretty sure it'll be fine as long as you dispose of data that's no longer needed properly (i.e. when the freemode ped is removed).
~~You'll need to store the heritage settings (all of them, parents and mix values)~~ "_GET_PED_HEAD_BLEND_DATA" was apparently figured out, you could test it, I have no idea how to pass the struct in C#, never really used structs. You'll also need to store facial features, overlay colors (only some overlays have colors, and you only need to store colors since you can get the current overlay with a native) and if you want to implement tattoos, then tattoos as well (tattoos have no slots or limits, you apply it to the ped, if you want to remove one, you clear them all and reapply everything but the one(s) you wish to remove).

If you're still against that, what would you think about adding it as an example code instead? I know I wanted to figure out how the online player models worked, pretty sure I'm not the only one.

File: source/scripting/Style.cs [hunk 1]
```diff
@@ -115,7 +115,7 @@ public IEnumerator<IPedVariation> GetEnumerator()
 			return (GetAllVariations() as IEnumerable<IPedVariation>).GetEnumerator();
 		}
 
-		public void RandmoizeOutfit()
+		public void RandmizeOutfit()
```
Comment (sollaholla) [inline source/scripting/Style.cs]: That's still not spelled correctly :P

**Comment by zorg93** [comment]
Are you wanting to pass a struct pointer to a native.
If so then you use [StructLayout(LayoutKind.Explicit)] attribute when creating the struct, Then each item you specify its offset in memory

A NativeVector3 would look like
'	[StructLayout(LayoutKind.Explicit)]
	struct NativeVector3
	{
		[FieldOffset(0x00)] float X; //first item
		[FieldOffset(0x08)] float Y; //second item(for native types items are stored 8 bytes apart
		[FieldOffset(0x10)] float Z; //third item as above
	}

	//then to call it in a native
	NativeVector3 temp;
	Function.Call(Hash.ABSF, new IntPtr(&temp));'

**Comment by DorCoMaNdO** [comment]
Thanks for explaining!
This worked:
```[StructLayout(LayoutKind.Explicit)]
        struct HeadBlendData
        {
            [FieldOffset(0x00)]
            public int shapeFirst;
            [FieldOffset(0x08)]
            public int shapeSecond;
            [FieldOffset(0x10)]
            public int shapeThird;
            [FieldOffset(0x18)]
            public int skinFirst;
            [FieldOffset(0x20)]
            public int skinSecond;
            [FieldOffset(0x28)]
            public int skinThird;
            [FieldOffset(0x30)]
            public float shapeMix;
            [FieldOffset(0x38)]
            public float skinMix;
            [FieldOffset(0x40)]
            public float thirdMix;
        }

        HeadBlendData data = new HeadBlendData();
        Function.Call<bool>(Hash._GET_PED_HEAD_BLEND_DATA, Game.Player.Character, new IntPtr(&data));

        new UIText("shapeFirst - " + data.shapeFirst.ToString(), new Point(200, 200), 0.25F).Draw();
        new UIText("shapeSecond - " + data.shapeSecond.ToString(), new Point(200, 220), 0.25F).Draw();
        new UIText("shapeThird - " + data.shapeThird.ToString(), new Point(200, 240), 0.25F).Draw();
        new UIText("skinFirst - " + data.skinFirst.ToString(), new Point(200, 260), 0.25F).Draw();
        new UIText("skinSecond - " + data.skinSecond.ToString(), new Point(200, 280), 0.25F).Draw();
        new UIText("skinThird - " + data.skinThird.ToString(), new Point(200, 300), 0.25F).Draw();
        new UIText("shapeMix - " + data.shapeMix.ToString(), new Point(200, 320), 0.25F).Draw();
        new UIText("skinMix - " + data.skinMix.ToString(), new Point(200, 340), 0.25F).Draw();
        new UIText("thirdMix - " + data.thirdMix.ToString(), new Point(200, 360), 0.25F).Draw();
```

Don't see any reason not to implement it if you can just use natives to get/set data.

**Comment by crosire** [comment]
I disabled unsafe code for the C# code. Everything unsafe should be done through the internal C++/Cli `MemoryAccess` class.

I do agree with @zorg93 that head morphs aren't necessarily something that is worth adding, since it doesn't apply to the default singleplayer characters. The idea of SHVDN is to provide general access to the scripting functionality of the game. Anything more sophisticated that is built on top of SHVDN's toolset should be put into a shared library instead (take NativeUI as an example). This would be a good candidate for one.

**Comment by DorCoMaNdO** [comment]
Well I have a character creator script that I built with a static API class for other scripts to use, when I started I was planning to release it, I'm not so sure whether I want to now, it's mainly due to how I've done it, when I worked with NativeUI months ago, whenever it was open, it caused my framerate to drop by 30-40 fps, on a GTX 980, so I don't trust it, instead I made a simple menu system that has the features I need, so if I'm going to release it I'll need to include my menu system, and it's quite barebones and I don't expect it to be well received cause of that.

I could send you the code privately if you'd like to provide some feedback (hasn't been through any cleaning though, and I made some changes to the API, some things that could make the code much shorter in some areas but I have yet to take advantage of those changes).

**Comment by zorg93** [comment]
The code for head blend most likely won't find its way in here though, there would be no harm releasing a small dll that people can reference for making head blends, you can even put it as an extension to the style class

**Comment by zorg93** [comment]
@DorCoMaNdO just realised an extension class wont be possible as the ped field in the style class is private

**Comment by DorCoMaNdO** [comment]
Create a public property with a private setter?

**Comment by zorg93** [comment]
Given that the only way to have a style class is via the Ped.Style property(due to the internal ctor), the only reason for exposing the ped field would be for this very reason. You may as well just keep it as an extension to the ped class

**Comment by DorCoMaNdO** [comment]
Technically I could still get it with reflection, but as far as I know that's a slow task, though only needed upon creation and the slower performance shouldn't be noticeable.
I'll look into that and other options once the PR is merged.

**Comment by DorCoMaNdO** [comment]
It's been a while since this discussion, my drive crashed some time ago and I lost many of my projects, including the character creator/API I was working on.
I've decompiled the latest binaries that I had shared with a friend and started recovering some code and a few days ago I noticed something strange, I ended up back at issue #521 due to the search terms I used, and this is probably not the best place to ask, but one of the recent updates to the game (I am assuming the latest one - v1.0.1290.1, as that's the only one I've noticed the issue on) seems to have broken the `_IS_PED_PROP_VALID` native, after setting 36 different ped props onto a ped (easy to do with a freemode model), `_IS_PED_PROP_VALID` always returns false for any of the ped props that you have yet to set (it would only return true for the 36 that you've used, not sure about whether texture variations affect this, can't recall), the only way to fix it is by using a different ped (or for the player, re-applying the model) but then it happens again 36 ped props in, I haven't played the game or messed around with ped props recently so I can't guarantee that it's the latest update to have caused this, but I can guarantee that it wasn't an issue a couple of months ago or when this discussion was ongoing.

Could anyone else try and reproduce this?

Also note that while `_IS_PED_PROP_VALID` returns false, if you know that a certain ped prop is valid, setting it still works. I'm not sure if there are any game crashing/breaking ped props like there are components but while testing it didn't seem like trying to set a ped's ped prop to an invalid index caused any issues, if there aren't then I think that making use of `GET_NUMBER_OF_PED_PROP_DRAWABLE_VARIATIONS` and `GET_NUMBER_OF_PED_PROP_TEXTURE_VARIATIONS` would work for an alternative, although I would like to know why `_IS_PED_PROP_VALID` breaks and know whether there's a (new?) alternative to replace it (I guess by checking decompiled scripts?) or if it's a SHV issue.

**Pull request merged**